middleware/basicauth/basicauth.go

```diff
@@ -31,6 +31,7 @@ func (a BasicAuth) ServeHTTP(w http.ResponseWriter, r *http.Request) (int, error
 			// Check credentials
 			if !ok || username != rule.Username || password != rule.Password {
 				w.Header().Set("WWW-Authenticate", "Basic")
+				w.WriteHeader(http.StatusUnauthorized)
```

Comment:
This is actually incorrect behavior and will result in multiple WriteHeader calls. See [this godoc](https://github.com/mholt/caddy/blob/b2ee6638e46973e6d0cf00ece0dbbc08947557fd/middleware/middleware.go#L22-L26) and [where the WriteHeader call actually takes place](https://github.com/mholt/caddy/blob/b2ee6638e46973e6d0cf00ece0dbbc08947557fd/server/server.go#L179-L182). Since this is a form of error handling, it is deferred to the application (or an error handling middleware, if the user chooses to) rather than being handled in the request handler.

It seems like a bug, and it's kind of a different way of thinking, but that is actually the correct way to do it.